### PR TITLE
Do not expect that Acquires will be matched by Release

### PR DESCRIPTION
**Pull request by rodwiddowson**
This is very grotty and edge case but kinda matters (and is also incredibly non-intuitive unless you think too har about the FltMgr implementation) but as a good rule you should not acquire any resources in the Acquire calls (which provoke many philosophical questions).

This is a bad initial attempt to explain this.  I'm sure Lori and Ted will do better, but that was 10 days of my life I wont get back debugging this so I wanted to capture it (and share my angst with the guys over in Redmond :-)

### CODE REVIEW

File: wdk-ddi-src/content/fltkernel/ns-fltkernel-_flt_parameters.md [hunk 1]
```diff
@@ -105,6 +105,11 @@ The following I/O operations do not have parameters, and therefore do not have a
 - IRP_MJ_SHUTDOWN
 - IRP_MJ_VOLUME_DISMOUNT
 
+For the operations that come in Acquire/Release pairs (ModWrite, CcFlush, Section), there is no guarantee that the release call will happen even after a sucessfull call to the post acquire callback.  An example would be a if an instance detach occurs.
+
+Counterinutitively, you should therefore avoid acquring resources in the Acquire call and releasing them in the release call.  If you have to you should either stall the instance detach (*before* Instance TearDownStart callback) or arrange for the resources to be released during instance tear down.
```
Comment (cgallred) [inline wdk-ddi-src/content/fltkernel/ns-fltkernel-_flt_parameters.md]: I would strike the second paragraph since this is true for any callback. It is a general rule that a filter should not acquire resources and then leave them held after exiting a callback. Doing what is described here is even worse than acquiring a resource in a pre-operation callback and releasing it in a post-operation callback. This describes getting the pre- and post-operation callbacks for a particular operation (e.g. `IRP_MJ_ACQUIRE_FOR_SECTION_SYNCHRONIZATION`) and then _leaving them held_ until some unknown future time.

File: wdk-ddi-src/content/fltkernel/ns-fltkernel-_flt_parameters.md [hunk 2]
```diff
@@ -105,6 +105,11 @@ The following I/O operations do not have parameters, and therefore do not have a
 - IRP_MJ_SHUTDOWN
 - IRP_MJ_VOLUME_DISMOUNT
 
+For the operations that come in Acquire/Release pairs (ModWrite, CcFlush, Section), there is no guarantee that the release call will happen even after a sucessfull call to the post acquire callback.  An example would be a if an instance detach occurs.
```
Comment (cgallred) [inline wdk-ddi-src/content/fltkernel/ns-fltkernel-_flt_parameters.md]: I suppose that being a Filter Manager owner means I think too hard about the FltMgr implementation, because this interpretation never occurred to me. :-) It is technically true that the operations that have "acquire" or "release" in their title come in pairs, but the same can be said of things like `IRP_MJ_CREATE`/`IRP_MJ_CLEANUP`, `IRP_MJ_PREPARE_MDL_WRITE`/`IRP_MJ_MDL_WRITE_COMPLETE` and maybe others. From a filter's perspective, just because certain operations a "paired" doesn't mean your filter will last long enough to see both operations.

Granted, the "acquire"/"release" operations normally happen closer together than e.g. create/cleanup, but the principle is the same. Given that, I'd reword this paragraph to be more of a reminder and clarification:

```suggestion
Note that the terms "acquire" or "release" in some operation names, such as `IRP_MJ_ACQUIRE_FOR_CC_FLUSH` and `IRP_MJ_RELEASE_FOR_CC_FLUSH`, do not imply any guarantees that a minifilter will always see both operations. Just as with operations such as `IRP_MJ_CREATE` and `IRP_MJ_CLEANUP`, a minifilter may be prevented from observing the second operation by other factors, such as receiving an _[InstanceTeardownStartCallback](/windows-hardware/drivers/ddi/fltkernel/nc-fltkernel-pflt_instance_teardown_callback)_ before the second operation occurs.
```

**Review by cgallred — CHANGES_REQUESTED** [review]

**Comment by lorihollasch** [comment]
Thanks for the PR, @rodwiddowson, and for the review and feedback @cgallred! I went ahead and updated the page to complete this PR.
